### ISSUE TITLE
LLVM to SMT: `<incomingBlock, currentBB> pair not found in EdgeAssertionsMap` on v6.5

### ISSUE THREAD

**Issue by pchaigno** [bpfverif/agni#2]
I'm hitting an error with the LLVM to SMT step, when running on Linux v6.5. All prior versions worked fine.
```
# clang --version
Ubuntu clang version 12.0.1-19ubuntu3
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
# mkdir ~/ebpf-verif-results-6.5 && python3 generate_encodings.py --kernver 6.5 --kernbasedir ~/linux-stable --outdir ~/ebpf-verif-results-6.5 --scriptsdir ~/ebpf-range-analysis-verification-cav23/llvm-to-smt/llvm-passes
Log file: ~/ebpf-verif-results-6.5/log_15_17_05_10_2023.log
Log error file: ~/ebpf-verif-results-6.5/log_err_15_17_05_10_2023.log
Change to kernel directory: ~/linux-stable ... done
Checkout kernel version v6.5 ... done
Run make config and edit BPF flags ... done
Extract compile flags for current kernel version ... done
Edit tnum.c and verifier.c to add wrappers ... done
Compile verifier.c and tnum.c ... done
Link verifier.ll and tnum.ll to single verifier.ll ... done
Getting encoding for BPF_ADD ... done
[...]
Getting encoding for BPF_XOR_32 ... done
Getting encoding for BPF_JEQTraceback (most recent call last):
  File "~/ebpf-range-analysis-verification-cav23/llvm-to-smt/generate_encodings.py", line 646, in <module>
    llvmpassrunner_for_op.run()
  File "~/ebpf-range-analysis-verification-cav23/llvm-to-smt/run_llvm_passes.py", line 160, in run
    self.run_llvm_to_smt_pass(self.inputdir_fullpath, "{}.ll".format(
  File "~/ebpf-range-analysis-verification-cav23/llvm-to-smt/run_llvm_passes.py", line 136, in run_llvm_to_smt_pass
    subprocess.run(cmd_llvm_to_smt_op, shell=True, check=True, text=True,
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '~/ebpf-range-analysis-verification-cav23/llvm-to-smt/llvm-passes/llvm_to_smt.sh ~/ebpf-verif-results-6.5 BPF_JEQ.ll check_cond_jmp_op_wrapper_BPF_JEQ 16 BPF_JEQ.smt2' returned non-zero exit status 1.

# ~/ebpf-range-analysis-verification-cav23/llvm-to-smt/llvm-passes/llvm_to_smt.sh ~/ebpf-verif-results-6.5 BPF_JEQ.ll check_cond_jmp_op_wrapper_BPF_JEQ 16 BPF_JEQ.smt2 &> llvm_to_smt.log
```

I'm attaching the full output here (zipped because it's too big for GitHub otherwise): [llvm_to_smt.zip](https://github.com/bpfverif/ebpf-range-analysis-verification-cav23/files/12820601/llvm_to_smt.zip)

I think the relevant part from that output is:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  <incomingBlock, currentBB> pair not found in EdgeAssertionsMap

PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /usr/bin/opt -load-pass-plugin ~/ebpf-range-analysis-verification-cav23/llvm-to-smt/llvm-passes/LLVMToSMT/build/libLLVMToSMT.so --passes=print<llvm-to-smt> --disable-output ~/ebpf-verif-results-6.5/BPF_JEQ.ll
 #0 0x00007f05d9a012a1 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xd732a1)
 #1 0x00007f05d99ff3d0 llvm::sys::RunSignalHandlers() (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xd713d0)
 #2 0x00007f05d9a019ad (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xd739ad)
 #3 0x00007f05d8775520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007f05d87c99fc pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x969fc)
 #5 0x00007f05d8775476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #6 0x00007f05d875b7f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #7 0x00007f05d8b04b9e (/lib/x86_64-linux-gnu/libstdc++.so.6+0xa2b9e)
 #8 0x00007f05d8b1020c (/lib/x86_64-linux-gnu/libstdc++.so.6+0xae20c)
 #9 0x00007f05d8b10277 (/lib/x86_64-linux-gnu/libstdc++.so.6+0xae277)
#10 0x00007f05d8b104d8 (/lib/x86_64-linux-gnu/libstdc++.so.6+0xae4d8)
#11 0x00007f05d65e39fa FunctionEncoder::handlePhiNode(llvm::PHINode&, int) ~/ebpf-range-analysis-verification-cav23/llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp:617:61
#12 0x00007f05d65e7a6c FunctionEncoder::encodeFunctionBody(llvm::Function&) ~/ebpf-range-analysis-verification-cav23/llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp:1283:27
#13 0x00007f05d65e9abd FunctionEncoder::buildSMT() ~/ebpf-range-analysis-verification-cav23/llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp:1572:54
#14 0x00007f05d6602e93 LLVMToSMT::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) ~/ebpf-range-analysis-verification-cav23/llvm-to-smt/llvm-passes/LLVMToSMT/LLVMToSMT.cpp:43:11
#15 0x00007f05d6612cbb llvm::detail::AnalysisPassModel<llvm::Module, LLVMToSMT, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Module>::Invalidator>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) /usr/include/llvm-12/llvm/IR/PassManagerInternal.h:315:65
#16 0x00007f05d9b677e2 llvm::AnalysisManager<llvm::Module>::getResultImpl(llvm::AnalysisKey*, llvm::Module&) (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xed97e2)
#17 0x00007f05d66080c2 LLVMToSMT::Result& llvm::AnalysisManager<llvm::Module>::getResult<LLVMToSMT>(llvm::Module&) /usr/include/llvm-12/llvm/IR/PassManager.h:789:22
#18 0x00007f05d66033ea LLVMToSMTPrinter::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) ~/ebpf-range-analysis-verification-cav23/llvm-to-smt/llvm-passes/LLVMToSMT/LLVMToSMT.cpp:91:50
#19 0x00007f05d6612daf llvm::detail::PassModel<llvm::Module, LLVMToSMTPrinter, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Module> >::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) /usr/include/llvm-12/llvm/IR/PassManagerInternal.h:86:3
#20 0x00007f05d9b647c2 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module> >::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xed67c2)
#21 0x00000000004243f0 llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::StringRef>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool) (/usr/bin/opt+0x4243f0)
#22 0x0000000000432874 main (/usr/bin/opt+0x432874)
#23 0x00007f05d875cd90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#24 0x00007f05d875ce40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#25 0x000000000041f9f5 _start (/usr/bin/opt+0x41f9f5)
~/ebpf-range-analysis-verification-cav23/llvm-to-smt/llvm-passes/llvm_to_smt.sh: line 50: 552689 Aborted                 (core dumped) /usr/bin/opt -load-pass-plugin ~/ebpf-range-analysis-verification-cav23/llvm-to-smt/llvm-passes/LLVMToSMT/build/libLLVMToSMT.so --passes="print<llvm-to-smt>" --disable-output ~/ebpf-verif-results-6.5/BPF_JEQ.ll
```

**Comment by pchaigno** [comment]
I bisected this to upstream commit [`51302c951c8f`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=51302c951c8fd5c298565c7127c855bf1d4550b6). That commit makes seemingly unrelated changes to the `reg_type_not_null` function (which becomes `reg_not_null`).

I then compared the LLVM IR of v6.4 with the same after this commit and noticed that it changed to now use a `switch` statement:
```llvm
if.end.i.i:                                       ; preds = %if.then.i
  %i.i.i.i = and i32 %i.i.i, 255
  %trunc = trunc i32 %i.i.i to i8
  switch i8 %trunc, label %lor.rhs.i.i [
    i8 13, label %reg_not_null.exit.i
    i8 12, label %reg_not_null.exit.i
    i8 11, label %reg_not_null.exit.i
    i8 5, label %reg_not_null.exit.i
    i8 4, label %reg_not_null.exit.i
    i8 16, label %land.lhs.true.i.i
  ]
```
(`if.end.i.i` is one of the basic blocks of `reg_not_null`.)

Then following the FunctionEncoder logic, it looks like we simply never process switch instructions (only branch instructions) and therefore didn't populate `EdgeAssertionsMap` (or `pathConditionsMap`) in preparation for the conversion to SMT.

I'm currently trying a fix.
